C# 6.0 feature: Exception Filters (#2)
dotnet/csharpstandard: standard/statements.md

- A `try` block followed by a `finally` block.

> - A `try` block followed by one or more `catch` blocks followed by a `finally` block.

Reviewer: This original para & bullet list are redundant, the forms are not referenced elsewhere and it simply a prose description of the `try_statement` production. In the original that production only has two alternatives:

```
try_statement
    : 'try' block catch_clauses
    | 'try' block catch_clauses* finally_clause
    ;
```

while the proposal expands this to three, maybe just to match this (original) bullet list, though the described syntax is identical.

The descriptions of other statements do not generally start with a prose description of the statement structure, so there is not a pattern being followed here, however the switch statement does.

Option 1: 😄 Remove this para & bullet list

Option 2: 🚀 Follow the model of the switch statement and replaced with a more succinct description which doesn't introduce "forms". Something like:

> A *try_statement* consists of the keyword `try` followed by a *block*, then zero or more *catch_clauses*, then an optional *finally_clause*. There must be at least one *catch_clause* or a *finally_clause*.

Option 3: 👀 Leave as is

If Option 1 or 2 is chosen then the grammar for *try_statement* can revert to the two alternative original version, of option 3 is chosen then there is a small argument for the proposed three alternative version.

> When a `catch` clause specifies an *exception_specifier*, the type shall be `System.Exception` or a type that derives from `System.Exception`.

Author: Done.

Unless a `catch` clause includes an exception variable name, it is impossible to access the exception object in the filter and `catch` block.

A `catch` clause that specifies neither an exception type nor an exception variable name is called a general `catch` clause. A `try` statement can only have one general `catch` clause, and, if one is present, it shall be the last `catch` clause.

> *Note*: Some programming languages might support exceptions that are not representable as an object derived from `System.Exception`, although such exceptions could never be generated by C# code. A general `catch` clause might be used to catch such exceptions. Thus, a general `catch` clause is semantically different from one that specifies the type `System.Exception`, in that the former might also catch exceptions from other languages. *end note*

> In order to locate a handler for an exception, `catch` clauses are examined in lexical order. If a `catch` clause specifies a type but no exception filter, it is a compile-time error for a later `catch` clause in the same `try` statement to specify a type that is the same as, or is derived from, that type. If a `catch` clause specifies no type and no filter, it shall be the last `catch` clause for that `try` statement.

Author: Line numbers moved. The above comment references 1122 and 1116 now instead of 1116 and 1112, respectively.  I made the edit.

- A `try` block followed by a `finally` block.
- A `try` block followed by one or more `catch` blocks followed by a `finally` block.

In an *exception_specifier* the *type*, or its effective base class if it is a *type_parameter*, shall be `System.Exception` or a type that derives from it.

> When a `catch` clause specifies both a *class_type* and an *Identifier*, an ***exception variable*** of the given name and type is declared. The exception variable corresponds to a local variable with a scope that extends over the `catch` block. During execution of the *exception_filter* and `catch` block, the exception variable represents the exception currently being handled. For purposes of definite assignment checking, the exception variable is considered definitely assigned in its entire scope.

Reviewer 2: It seems odd that the exception is already "being handled" during execution of the *exception_filter* even though an "exception handler" has not been located yet.

Unless a `catch` clause includes an exception variable name, it is impossible to access the exception object in the filter and `catch` block.

A `catch` clause that specifies neither an exception type nor an exception variable name is called a general `catch` clause. A `try` statement can only have one general `catch` clause, and, if one is present, it shall be the last `catch` clause.

> *Note*: Some programming languages might support exceptions that are not representable as an object derived from `System.Exception`, although such exceptions could never be generated by C# code. A general `catch` clause might be used to catch such exceptions. Thus, a general `catch` clause is semantically different from one that specifies the type `System.Exception`, in that the former might also catch exceptions from other languages. *end note*

> In order to locate a handler for an exception, `catch` clauses are examined in lexical order. If a `catch` clause specifies a type but no exception filter, it is a compile-time error for a later `catch` clause in the same `try` statement to specify a type that is the same as, or is derived from, that type.

Reviewer 3: ```suggestion
In order to locate a handler for an exception, `catch` clauses are examined in lexical order. If a `catch` clause specifies a type but no exception filter, it is a compile-time error for a later `catch` clause of the same `try` statement to specify a type that is the same as, or is derived from, that type.
```

  - If the `try` statement has a `finally` block, the `finally` block is executed.
  - Control is transferred to the end point of the `try` statement.
- If an exception is propagated to the `try` statement during execution of the `try` block:

>   - The `catch` clauses, if any, are examined in order of appearance to locate a suitable handler for the exception. If a `catch` clause does not specify a type, or specifies the exception type or a base type of the exception type is considered a match. A general `catch` clause is considered a match for any exception type. If a matching `catch` clause is located:

Reviewer 3: ```suggestion
  - The `catch` clauses, if any, are examined in order of appearance to locate a suitable handler for the exception. Each `catch` clause that does not specify a type, or specifies the exception type or a base type of the exception type, is considered a match. If a matching `catch` clause is located:
```

Reviewer 3: Remove duplicate description of general catch clause.

    - If the matching `catch` clause declares an exception variable, the exception object is assigned to the exception variable.

>     - If the `catch` clause declares an exception filter, the filter is evaluated. If it evaluates to `false`, the catch clause is not a match, and the search continues through any subsequent `catch` clauses for a suitable handler.    

Reviewer 3: ```suggestion
    - If the `catch` clause declares an exception filter, the filter is evaluated. If it evaluates to `false`, the search continues through any subsequent `catch` clauses for a suitable handler.    
```

Reviewer 3: Remove text that asserts that a catch clause that is a match is not a match.

>     - Otherwise, the `catch` clause is considered a match, and control is transferred to the matching `catch` block.

Reviewer 3: ```suggestion
    - Otherwise control is transferred to the corresponding `catch` block.
```
